server/routes.go

```diff
@@ -168,93 +193,154 @@ func GenerateHandler(c *gin.Context) {
-			if r.Done {
-				r.TotalDuration = time.Since(checkpointStart)
-				r.LoadDuration = checkpointLoaded.Sub(checkpointStart)
-			}
+func ChatHandler(c *gin.Context) {
+	loaded.mu.Lock()
+	defer loaded.mu.Unlock()
 
-			if req.Raw {
-				// in raw mode the client must manage history on their own
-				r.Context = nil
-			}
+	checkpointStart := time.Now()
 
-			ch <- r
-		}
+	var req api.ChatRequest
+	err := c.ShouldBindJSON(&req)
+	switch {
+	case errors.Is(err, io.EOF):
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "missing request body"})
+		return
+	case err != nil:
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
 
-		if err := loaded.runner.Predict(c.Request.Context(), req.Context, prompt, req.Format, fn); err != nil {
-			ch <- gin.H{"error": err.Error()}
-		}
-	}()
+	// validate the request
+	switch {
+	case req.Model == "":
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "model is required"})
+		return
+	case len(req.Format) > 0 && req.Format != "json":
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "format must be json"})
+		return
+	}
+
+	sessionDuration := defaultSessionDuration
+	model, err := fetchAndLoadModel(c, req.Model, req.Options, sessionDuration)
+	if err != nil {
+		// error set in request context by fetchAndLoadModel
+		return
+	}
+
+	// an empty request loads the model
+	if len(req.Messages) == 0 {
+		c.JSON(http.StatusOK, api.PredictResponse{CreatedAt: time.Now().UTC(), Model: req.Model, Done: true})
+		return
+	}
+
+	checkpointLoaded := time.Now()
+
+	prompt, err := promptFromMessages(model, req.Messages)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
+
+	predict := Predict{
+		Model:            model,
+		Prompt:           prompt,
+		Format:           req.Format,
+		CheckpointStart:  checkpointStart,
+		CheckpointLoaded: checkpointLoaded,
+		SessionDuration:  sessionDuration,
+		ResponseCallback: func(r *api.PredictResponse) {
+			if len(req.Messages) > 0 && !r.Done {
+				r.Message = &api.Message{Role: "assistant", Content: r.Response}
+				// Do not send back response in the case of messages
+				r.Response = ""
+			}
+		},
+	}
+	ch, generated := predict.Run(c.Request.Context())
 
 	if req.Stream != nil && !*req.Stream {
-		var response api.GenerateResponse
-		generated := ""
+		// Wait for the channel to close
+		var r api.PredictResponse
 		for resp := range ch {
-			if r, ok := resp.(api.GenerateResponse); ok {
-				generated += r.Response
-				response = r
-			} else {
+			var ok bool
+			if r, ok = resp.(api.PredictResponse); !ok {
 				c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
 				return
 			}
 		}
-		response.Response = generated
-		c.JSON(http.StatusOK, response)
+		if len(req.Messages) > 0 {
+			r.Message = &api.Message{Role: "assistant", Content: generated.String()}
+		} else {
+			r.Response = generated.String()
```

Comment:
I don't like how this changes the the response format – `/api/chat` should not respond with `response`

We can either return an empty message object or nil (`message` field omitted)

Reply:
Updated to return earlier in the chat/generate logic if the prompt/messages are empty (before any generation happens)